jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -72,34 +71,25 @@ public Class findClass(String name) {
         .filter(path -> path.toString().endsWith(".class"))
         .walk(
             classFile -> {
-              // Convert filename (rootDir/path/to/ClassName.class) to class name
-              // (path.to.ClassName)
-              Path relativized = rootDirectory.relativize(classFile);
-              StringJoiner stringJoiner = new StringJoiner(".");
-              for (Path path : relativized) {
-                stringJoiner.add(path.toString());
+              Class<?> fileClass = new ClassFileLoader(classFile).findClass(null);
+              if (fileClass == null) {
+                return;
               }
-              String name = stringJoiner.toString();
-              name = name.substring(0, name.length() - ".class".length());
-
-              Class<?> fileClass = new ClassFileLoader(classFile).findClass(name);
-              if (fileClass != null) {
-                try {
-                  // Check if class contains {@code public static void main(String[] args)}
-                  Method main = fileClass.getMethod("main", String[].class);
-                  if (main != null
-                      && main.getReturnType() == void.class
-                      && Modifier.isStatic(main.getModifiers())
-                      && Modifier.isPublic(main.getModifiers())) {
-                    results.add(name);
-                  }
-                } catch (NoSuchMethodException ignored) {
-                  // main method not found
+              try {
+                // Check if class contains {@code public static void main(String[] args)}
+                Method main = fileClass.getMethod("main", String[].class);
+                if (main != null
+                    && main.getReturnType() == void.class
+                    && Modifier.isStatic(main.getModifiers())
+                    && Modifier.isPublic(main.getModifiers())) {
+                  classNames.add(fileClass.getCanonicalName());
```

Comment:
I believe you would want `getName` since the canonical name would replace a subclass like `package.Hello$Main` with `package.Hello.Main` 

https://stackoverflow.com/questions/15202997/what-is-the-difference-between-canonical-name-simple-name-and-class-name-in-jav

Reply:
K, wasn't sure which one we wanted.